src/render.rs

```diff
@@ -373,32 +379,48 @@ impl<'a> Renderer<'a> {
                 self.render_for(variable, array, body)
             },
             Block {name, body} => {
-                match self.template.blocks.get(&name) {
-                    Some(b) => {
-                        match b.clone() {
-                            Block {body, ..} => {
-                                self.render_node(*body.clone())
-                            },
-                            _ => unreachable!()
-                        }
-                    },
-                    None => {
-                        self.render_node(*body)
-                    }
+                self.block_definitions.push((self.templates.last().unwrap(), Block {name: name.clone(), body: body.clone()}));
+                self.block_definitions.extend(self.templates.iter().rev().skip(1).flat_map(|template| match template.blocks.get(&name) {
+                    Some(block @ &Block {..}) => Some((*template, block.clone())),
+                    Some(_) => unreachable!(),
+                    None => None
+                }));
+                self.block_stack_idx = self.block_definitions.len() - 1;
+                let body = match self.block_definitions[self.block_stack_idx].1 {
+                    Block {ref body, ..} => *body.clone(),
+                    _ => unreachable!()
+                };
+                self.render_node(body)
+            },
+            Super => {
+                if self.block_definitions.is_empty() {
+                    Err(SuperOutsideBlock(self.templates.last().unwrap().name.clone()))
+                } else if self.block_stack_idx == 0 {
+                    let template_name = self.block_definitions[self.block_stack_idx].0.name.clone();
+                    let block_name = match &self.block_definitions[self.block_stack_idx].1 {
+                        &Block {ref name, ..} => name.clone(),
+                        _ => unreachable!()
+                    };
+                    Err(SuperNoParent(template_name, block_name))
```

Comment:
that can be moved at parsing time as well, when we insert the blocks in `Template`, we already know if we have a parent or not